### ISSUE TITLE
Method to retrieve weights as a DataArray

### ISSUE THREAD

**Issue by aulemahal** [pangeo-data/xESMF#27]
While coding the spatial averaging in #24, I realized that the Regridder object was missing an option to retrieve a usable array of the regridding weights. While this is not common when performing conventional regridding, I think that most users doing spatial averaging would like to see or save the weights masks.

The `Regridder` object could have a simple `get_weights` method that returns a DataArray of the expanded sparse matrix with 4 or 3 dimensions. Dimensions would be renamed with "_in" and "_out" suffixes.

**Comment by raphaeldussin** [comment]
my understanding is that ESMF only produces 2d regridding matrices (I could be wrong). 
The size of the sparse matrix being (nx_src * ny_src) x (nx_dst * ny_dst), are you suggesting something like:

```python
def get_weights(regridder):
    return xr.DataArray(regridder.weights.toarray(), dims=('nxy_src', 'nxy_src'))
```

**Comment by aulemahal** [comment]
I would suggest something a bit more complex that outputs an array of shape `(lon_in, lat_in, lon_out, lat_out)` or `(lon_in, lat_in, locations)` in a locstream/polylist_out case.

**Comment by huard** [comment]
@aulemahal Could you repurpose or generalize `smm.read_weights` ?

**Comment by aulemahal** [comment]
I'm not sure how that would serve the purpose. May be the issue is not clear, I am suggesting pretty much this:
```python3
def get_weights(self):
    if self.locstream_in:
        dims = ['locations_in']
        shape = [self.n_in]
    else:
        dims = [dim + '_in' for dim in self.input_horiz_dims]
         shape = list(self.shape_in)

    if self.locstream_out:
        dims.append('locations_out)]
        shape.append(self.n_out)
    else:
        dims.extend([dim  + '_out' for dim in self.out_horiz_dims])]
         shape.extend(self.shape_out)
    return xr.DataArray(self.weights.toarray().reshape(shape), dims=dims)
```
This would also need the new attribute `input_horiz_dims` to be set in the `__init__`, it could still be overridden in the regrid methods anyway.

**Comment by raphaeldussin** [comment]
@aulemahal I'm not sure that would work for large arrays. Say if I regrid a 1/4 degree regular grid onto a 1 degree (pretty standard stuff), a 4d matrix would be of size 1440x720x360x180 = 67e9 elements = 269 GB for single precision

**Comment by aulemahal** [comment]
You're right... I was thinking of the cases using polygons, where the weights could actually be useful themselves. This could be a small example in the doc instead. So, if the user explodes their ram, it's their own fault.

**Comment by aulemahal** [comment]
Just remembered this issue. It was "solved" by an example in the doc of #24.
